include/networkit/distance/AStarGeneral.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * AStarGeneral.hpp
+ *
+ *  Created on: 09.08.2019
+ *      Author: Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+#ifndef A_STAR_GENERAL_H
+#define A_STAR_GENERAL_H
+
+#include <networkit/distance/STSP.hpp>
+
+#include <tlx/container/d_ary_addressable_int_heap.hpp>
+
+namespace NetworKit {
+
+/**
+ * @ingroup distance
+ * AStar path-finding algoritm with general Heuristic function.
+ */
+template <typename Heuristic> class AStarGeneral final : public STSP {
+public:
+    /**
+     * Creates the AStarGeneral class for the graph @a G, the source node @a
+     * source, and the target node @a target using a template parameter as
+     * heuristic function.
+     *
+     * @param G The graph.
+     * @param heu Heuristic function that takes a node as input and returns a
+     * lower bound of the distance of that node to the target node.
+     * @param source The source node.
+     * @param target The target node.
+     * @param storePred If true, the algorithm will also store the predecessors
+     * and reconstruct a shortest path from @a source and @a target.
+     */
+    AStarGeneral(const Graph &G, Heuristic heu_, node source, node target, bool storePred = true)
+        : STSP(G, source, target, storePred), heu(heu_), heap(Compare(prio)) {}
+
+    /*
+     * Executes the AStar algorithm.
+     */
+    void run() override {
+        if (source == target) {
+            hasRun = true;
+            return;
+        }
+
+        init();
+        const edgeweight infdist = std::numeric_limits<edgeweight>::max();
```

Comment:
can be constexpr; type is specified redudantly. consider `constexpr auto infdist= ...`